src/lsm/composite_key.zig

```diff
@@ -4,67 +4,73 @@ const math = std.math;
 
 const stdx = @import("../stdx.zig");
 
-pub fn CompositeKey(comptime Field: type) type {
-    assert(Field == u128 or Field == u64);
+pub fn CompositeKeyType(comptime Field: type) type {
+    // The type if zeroed padding is needed.
+    const pad = switch (Field) {
+        u128 => @as(u64, 0),
+        u64 => {},
+        else => @compileError("invalid Field for CompositeKey: " ++ @typeName(Field)),
+    };
+    const Pad = @TypeOf(pad);
+
+    // The composite key's backing integer type.
+    const Int = std.meta.Int(
+        .unsigned,
+        @bitSizeOf(u64) + @bitSizeOf(Field) + @bitSizeOf(Pad),
+    );
 
-    return extern struct {
+    return packed struct(Int) {
         const Self = @This();
 
-        pub const sentinel_key: Self = .{
+        pub const sentinel_key: Key = @bitCast(Self{
             .field = math.maxInt(Field),
             .timestamp = math.maxInt(u64),
-        };
+        });
 
         const tombstone_bit = 1 << 63;
 
-        // The type if zeroed padding is needed after the timestamp field.
-        const pad = switch (Field) {
-            u128 => @as(u64, 0),
-            u64 => [0]u8{},
-            else => @compileError("invalid Field for CompositeKey: " ++ @typeName(Field)),
-        };
-
-        // u128 may be aligned to 8 instead of the expected 16.
-        const field_bitsize_alignment = @divExact(@bitSizeOf(Field), 8);
+        pub const Key = Int;
 
-        pub const Value = Self;
+        // The keys are ordered by (field, timestamp),
+        // fields are declared from the least siginificant to the most significant:
 
-        field: Field align(field_bitsize_alignment),
         /// The most significant bit must be unset as it is used to indicate a tombstone.
         timestamp: u64,
-        /// [0]u8 as zero-sized-type workaround for https://github.com/ziglang/zig/issues/16394.
-        padding: @TypeOf(pad) = pad,
+
+        field: Field,
+
+        padding: Pad = pad,
```

Comment:
:facepalm: I think I mixed things up when I said that padding first breaks alignment. I think we actually _need_ to have padding first to preserve alignment, _because_ timestamp precedes Field. 

For `Field = u128, timestamp = u64, pad = u64`, the traditional layout would be:

```
Field     // u128
timestamp // u64
Pad       // u64
```

This layout is well-aligned, but doesn't fit little-endian comparison order we need. 

With timestamp first,  we need this layout for alignment to add up:

```
Pad       // u64
timestamp // u64
Field     // u128
```


:thinking: honestly, I think it doesn't really matter either way. 

To confuse us with even more options: I realized that we don't actually need to change composite key layout here. We are doing non-trivial masking operation in `key_from_value` anyway, and that can read like:

```
return @as(Key, value.timestamp & ~tombstone_bin) | (@as(Key, value.field) << 64)
```

This then would be completely independent on the actual layout of data on disk.

I think I am slightly in favor of this latter option -- what makes me slightly uneasy about the current state in the PR, is that our "integer keys" affects the data layout on disk somewhat (as we rearrange the order of bytes), but it is _still_ not just "take these bytes as a key", as we need to mask away the tombstone. 

But we actually don't have to change the layout on disk! `fn key_from_value` allows us to cook up the integer key in whichever way we want. So we preserve both:

* the natural, aligned `field, timestamp, padding` layout on disk
* the integer keys

I don't have a _strong_ argument for why "natural" layout is actually good though, except that it is, well, non-weird :) 

Sorry for back and forth here!

Reply:
> return @as(Key, value.timestamp & ~tombstone_bin) | (@as(Key, value.field) << 64)

That's a great solution. We don't particularly care about keeping the data layout on disk as a little-endian integer, in fact, I think it gives more flexibility to the LSM layout.

> Sorry for back and forth here!
Not at all!!!